Dev/vxi11 adapter

chweiser opened the pull request:
This pull request will add a vxi11 adapter that resolves the #162 "Connect over Ethernet bridge E5810x ?".

bilderbuchi commented on pymeasure/adapters/vxi11.py:
```diff
@@ -0,0 +1,113 @@
+#
+
+from .adapter import Adapter
+import vxi11
```
This adds a dependency on the `python-vxi11` package, right?

bilderbuchi replied:
Also, if this becomes a default dependency, I think python-vxi11 then needs to be added to conda-forge, at least I couldn't find it there.

chweiser replied:
You are right. So I will go ahead and add python-vxi11 to the install_requirements section of the setup.py. Should i also add the requirement to all files within the  requirements folder?

cjermain commented on requirements/conda-3.4.yml:
```diff
@@ -13,6 +13,7 @@ dependencies:
 - pytest=2.9.2=py34_0
 - python=3.4.5=0
 - python-dateutil=2.5.3=py34_0
+- python-vxi11=0.9=py34_0
```
As far as I can tell, this package is only available in pip (https://pypi.org/project/python-vxi11/). Therefore this should go in the `pip` section at the bottom. Please make the change for all of the conda environment files.

bilderbuchi replied:
We should probably think about creating a conda forge feedstock for python-xvi11, otherwise pymeasure ceases to be installable via conda. maybe its maintainer is also interested in this?

cjermain commented on pymeasure/adapters/vxi11.py:
```diff
@@ -0,0 +1,113 @@
+
+
+    def __repr__(self):
+        return "VXI11Adapter connecting to: <{}>".format(self.connection.host)
```
Using `'<VXI11Adapter(host={})>'.format(self.connection.host)` is more consistent with the `__repr__` methods in the rest of the codebase.

cjermain commented on pymeasure/adapters/vxi11.py:
```diff
@@ -0,0 +1,113 @@
+               self.conn_kwargs[key] = kwargs[key]
+
+        self.connection = vxi11.Instrument(host, **self.conn_kwargs)
+
```
Only 1 space is needed between class methods.

chweiser replied:
changed in the new commit

cjermain commented:
@chweiser, thanks for putting together this PR! I'm excited to see VXI-11 support in PyMeasure, and will wait to get this in before putting together the 0.6 release.

In order to provide documentation for this new Adapter, we need to add it to the [adapter.rst file](https://github.com/ralph-group/pymeasure/blob/master/docs/api/adapters.rst).

bilderbuchi commented:
Does this need tests? Is this code something that is testable?

chweiser commented:
I have now changed the requirements, so that the python-vxi11 is only an extra requirement. 
Then i changed the object representation, as suggested and added the adapter to adapters.rst. I also reformatted the code to match the project standard. Please let me know if this is all in order now.

cjermain commented on pymeasure/adapters/__init__.py:
```diff
@@ -36,5 +36,6 @@
 try:
     from pymeasure.adapters.serial import SerialAdapter
     from pymeasure.adapters.prologix import PrologixAdapter
+    from pymeasure.adapters.vxi11 import VXI11Adapter
```
This should go in it's own try/catch clause to provide a relevant message. The PrologixAdapter is in there because it also requires PySerial.

cjermain commented:
@chweiser, nice work! For now, keeping python-vxi11 as an optional requirement sounds good. How will people know from the docs how to install the optional dependency?

@bilderbuchi, thanks for bringing up testing. Ideally all code should be tested in the package. We could mock the Instrument construction and test each method. Given the simplicity of the methods,. I am more concerned with testing the adapter on real hardware, and making sure that people can install the appropriate packages easily. I also appreciate your suggestion on conda-forge.

@MadaniJo, you mentioned being up to test this adapter. Can you test it on an instrument?

MadaniJo commented:
Sure will give it a try within this week.

On Thu, Oct 25, 2018, 08:37 Colin Jermain <notifications@github.com> wrote:

> @chweiser <https://github.com/chweiser>, nice work! For now, keeping
> python-vxi11 as an optional requirement sounds good. How will people know
> from the docs how to install the optional dependency?
>
> @bilderbuchi <https://github.com/bilderbuchi>, thanks for bringing up
> testing. Ideally all code should be tested in the package. We could mock
> the Instrument construction and test each method. Given the simplicity of
> the methods,. I am more concerned with testing the adapter on real
> hardware, and making sure that people can install the appropriate packages
> easily. I also appreciate your suggestion on conda-forge.
>
> @MadaniJo <https://github.com/MadaniJo>, you mentioned being up to test
> this adapter. Can you test it on an instrument?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ralph-group/pymeasure/pull/166#issuecomment-432873721>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ag8A-l9VHuBhfJF4HnfTN58TTOtYveVQks5uoQevgaJpZM4X1cxC>
> .
>

chweiser commented:
I absolutely encourage you to test  the adapter on real hardware. I have so far tested the adapter for Keysight 34400 Series DMMs, Keysight 33000 Series AWGs and a another instrument through an Agilent E5810A LAN  GPIB bridge. 
Regarding the documentation and the package installation I'm unsure and would appreciate any input.

chweiser commented:
Its seems like that you made your own version of the project, could you try to install the pull request in its own enviroment using:
`pip install git+https://github.com/ralph-group/pymeasure.git@refs/pull/166/head` 
followed by 
`pip install python-vxi11` 


And then run the following two tests. Remember to adjust change the ip adress and the gpib address.

Test 1 simply connect  to any instrument  using the vxi11 adapter

```
from pymeasure.adapters.vxi11 import VXI11Adapter
from pymeasure.instruments import Instrument

adapter = VXI11Adapter("TCPIP::10.45.26.139::inst0::INSTR")
instr = Instrument(adapter, "my_instrument")
print(instr.ask("*IDN?")) 
```

Test 2 connect to a device using a lan to gpib bridge.

```
from pymeasure.adapters.vxi11 import VXI11Adapter
from pymeasure.instruments import Instrument

adapter = VXI11Adapter("TCPIP0::192.168.1.159::gpib0,2:INSTR")
instr = Instrument(adapter, "my_instrument")
print(instr.ask("*IDN?"))
```

MadaniJo commented:
> 
> 
> Its seems like that you made your own version of the project, could you try to install the pull request in its own enviroment using:
> `pip install git+https://github.com/ralph-group/pymeasure.git@refs/pull/166/head`
> followed by
> `pip install python-vxi11`
> 
> And then run the following two tests. Remember to adjust change the ip adress and the gpib address.
> 
> Test 1 simply connect to any instrument using the vxi11 adapter
> 
> ```
> from pymeasure.adapters.vxi11 import VXI11Adapter
> from pymeasure.instruments import Instrument
> 
> adapter = VXI11Adapter("TCPIP::10.45.26.139::inst0::INSTR")
> instr = Instrument(adapter, "my_instrument")
> print(instr.ask("*IDN?")) 
> ```
> 
> Test 2 connect to a device using a lan to gpib bridge.
> 
> ```
> from pymeasure.adapters.vxi11 import VXI11Adapter
> from pymeasure.instruments import Instrument
> 
> adapter = VXI11Adapter("TCPIP0::192.168.1.159::gpib0,2:INSTR")
> instr = Instrument(adapter, "my_instrument")
> print(instr.ask("*IDN?"))
> ```

Hi Chweiser,
I still don't succeed to make it work on my side, may be I miss something stupid. Could you provide your test code so I can understand what I am doing wrong. At least the basic *IDN? command should work from there on I can the dig and modify what needs to be done. Thanks, Dan

chweiser commented:
@MadaniJo I am not entirely sure whats the problem here. Are you sure that you pass the correct connection string to the adapter? I can definitely connect to the instrument through the gpib bridge using the example code provided.
Pay attention to the ::gpib0,2: part of the string. Try to use your browser to connect to the brigde, then use its built in find feature to see the exact location of your instrument on the bus.

MadaniJo commented:
Dear Chweiser,

Connection over web interface work and VXI11 too. So do you mind to send me
a small snipset of your code *INI? part  would be good enough.
Anyway I will try your suggestion.

Regards,
Dan

Le lun. 26 nov. 2018 à 16:44, chweiser <notifications@github.com> a écrit :

> @MadaniJo <https://github.com/MadaniJo> I am not entirely sure whats the
> problem here. Are you sure that you pass the correct connection string to
> the adapter? I can definitely connect to the instrument through the gpib
> bridge using the example code provided.
> Pay attention to the ::gpib0,2: part of the string. Try to use your
> browser to connect to the brigde, then use its built in find feature to see
> the exact location of your instrument on the bus.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ralph-group/pymeasure/pull/166#issuecomment-441559983>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ag8A-levV18uaCgw6_ZRcX-GJ7UIRjrUks5uy6n1gaJpZM4X1cxC>
> .
>

MadaniJo commented:
Dear Chweiser,
High five man :). I must have missed some (:) somewhere. So it definitely
works.
My morning is starting good.
Thanks again.
Dan

Le lun. 26 nov. 2018 à 17:25, chweiser <notifications@github.com> a écrit :

> The code im running is basically what i posted earlier.
>
> from pymeasure.adapters.vxi11 import VXI11Adapter
> from pymeasure.instruments import Instrument
>
> adapter = VXI11Adapter("TCPIP::10.45.26.137::gpib0,17::INSTR")
> instr = Instrument(adapter, "my_instrument")
> print(instr.ask("*IDN?"))
>
> For me it returns the instrument identification string as expected.
>
> Try to post the error, if it still fails for you.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ralph-group/pymeasure/pull/166#issuecomment-441571663>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ag8A-hbIR-3tJaMrjNbqPWDzx1daO1IYks5uy7OWgaJpZM4X1cxC>
> .
>

cjermain commented:
@MadaniJo, thanks for testing this feature.

@chweiser, I think it is a great idea to add more documentation to the tutorial. I will go ahead and merge this PR -- please open another one of the documentation.
